core/src/main/java/org/elasticsearch/plugins/PluginsMetaDataUpgrader.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugins;
+
+import org.elasticsearch.cluster.metadata.MetaData;
+
+import java.util.Collection;
+import java.util.function.Function;
+
+/**
+ * Upgrades {@link MetaData} on startup on behalf of installed {@link Plugin}s
+ */
+public class PluginsMetaDataUpgrader {
```

Comment:
I would remove the "plugins" part from this, and have MetaDataUpgraders (which is only a container here because we cannot yet construct the GatewayMetaService in Node, and guice requires a unique type). I would just have a public final immutable collection as the only member, no methods.